### PR TITLE
Multioutput reduction strategy for forecasting

### PR DESCRIPTION
**Pull request by Lovkush-A**
#### Reference Issues/PRs

See #403 

#### What does this implement/fix? Explain your changes.

Created MultioutputRegressionForecaster class via a _MultioutputReducer base class. Almost identical to Direct strategy.

#### What should a reviewer concentrate their feedback on?

- [ ] Have I understood Multitoutput reduction?
- [ ] Is the test sufficient?
- [ ] Do I need to modify any other files?
- [ ] Should I also have Multioutput for Time series?
- [ ]  Does this count as a new estimator?

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/alan-turing-institute/sktime/blob/master/.all-contributorsrc).
- [ ] Optionally, I've updated sktime's [CODEOWNERS](https://github.com/alan-turing-institute/sktime/blob/master/CODEOWNERS) to receive notifications about future changes to these files.
- [ ] I've added unit tests and made sure they pass locally.

##### For new estimators
- [ ] I've added the estimator to the online documentation.
- [ ] I've updated the existing example notebooks or provided a new one to showcase how my estimator works.

### CODE REVIEW

**Comment by mloning** [comment]
Thanks @Lovkush-A - looks very good, I'll take a closer look in the next few days!

File: sktime/forecasting/compose/_reduce.py
Comment (mloning) [inline sktime/forecasting/compose/_reduce.py:288]:
```suggestion
            raise NotImplementedError("Exogenous variables `X` are not yet supported.")
```

File: sktime/forecasting/compose/_reduce.py
Comment (mloning) [inline sktime/forecasting/compose/_reduce.py:310]: What's the error here if the the regressor does not support multi-output? Is it user friendly or should we include some kind of check on the regressor if it is compatible with the multi output strategy?

**Review by mloning — APPROVED** [review]
Thanks @Lovkush-A looks all good to me! 

Only thing that we still have to do is to add it to the docs at `docs/source/api_reference/forecasting.rst`. 

Will merge it once that's done.

File: sktime/forecasting/compose/_reduce.py
Comment (Lovkush-A) [reply sktime/forecasting/compose/_reduce.py:310]: I do not know, nor do I know how I would find out. I am trusting the error is user-friendly...

1. How do I find a regressor that does not support multi-output? I searched on google and somebody asked and got no reponse. https://datascience.stackexchange.com/questions/25829/scikit-learn-which-regressors-natively-support-multi-target-regression
2. How would we check if a regressor is compatible?

File: sktime/forecasting/compose/_reduce.py
Comment (Lovkush-A) [reply sktime/forecasting/compose/_reduce.py:288]: Done (also in other parts of the file)

**Comment by Lovkush-A** [comment]
Should I add multitoutput for time series too?

**Comment by Lovkush-A** [comment]
Should I add multitoutput for time series too?

File: sktime/forecasting/compose/_reduce.py
Comment (mloning) [reply sktime/forecasting/compose/_reduce.py]: I think ultimately we can only have heuristics for finding out whether an estimator supports multiouput or not. One thing to look for is the multi output mixin class or multiouput tag from scikit-learn: https://github.com/scikit-learn/scikit-learn/blob/95119c13af77c76e150b753485c662b7c52a41a2/sklearn/base.py#L758

**Comment by mloning** [comment]
Thanks again @Lovkush-A, really appreciate your work on this! I think that adding support for exogenous variables is more important at the moment, leaving multioutput time series regression as a good first issue for other perhaps

File: sktime/forecasting/compose/_reduce.py
Comment (fkiraly) [reply sktime/forecasting/compose/_reduce.py]: to add a detail: if a `sklearn` estimator does not support multioutput, you can always make it support it by wrapping it in `MultiOutputRegressor` or `MultiOutputClassifier` (a kind of reduction/composition).

Comment (Lovkush-A) [reply sktime/forecasting/compose/_reduce.py]: Responding to the detail, if they estimator does not support multioutput, they should just use direct!

**Comment by Lovkush-A** [comment]
Alright, works for me. I will go onto exogenous next.

File: sktime/forecasting/compose/_reduce.py
Comment (mloning) [reply sktime/forecasting/compose/_reduce.py]: @Lovkush-A yes, wrapping in `MultiOutputRegressor` is the same as using `direct` - perhaps something we should add to the docstring. Similarly, there is [`RegressorChain`](https://scikit-learn.org/stable/modules/generated/sklearn.multioutput.RegressorChain.html) for recursive (or direct?) strategy.

Comment (Lovkush-A) [reply sktime/forecasting/compose/_reduce.py]: I am not sure if docstring is correct place for that. Feels like the kind of thing that should go in the user-guide.

Comment (fkiraly) [reply sktime/forecasting/compose/_reduce.py]: hm, this starts souding like an inconsistency!

I.e., if the wrapped estimator supports multi-output that is not identical with looping over output variables, what should the `direct` strategy do?

I think there are actually two direct strategies!
* `direct` - loops over the forecasting horizon, using single-output versions of the wrapped estimator; fits multiple single-output versions of the estimator internally
* `direct-multioutput` - uses the wrapped estimator's multi-output mode; fits a single multi-output estimator internally

While `direct-multioutput` after applying `MultiOutputRegressor` is always the same as `direct`, it is not true that `direct-multioutput` is always the same as `direct`, hence these are two different strategies for reduction!

Comment (fkiraly) [reply sktime/forecasting/compose/_reduce.py]: @mloning implementing it this way, i.e., properly splitting up the two reduction strategies here, would also remove the need to check whether the estimator supports multi-output. The docstring can say that multi-output is assumed to be supported if you use `direct-multioutput` as your reduction strategy.

Comment (Lovkush-A) [reply sktime/forecasting/compose/_reduce.py]: I do not follow. Your `direct-multioutput` sounds like `multioutput`.

Comment (fkiraly) [reply sktime/forecasting/compose/_reduce.py]: ok, I'm confused now, probably entirely my fault ... what are the current options for strategies?
And which option did your original comment refer to ("if not support multioutput")?

Comment (Lovkush-A) [reply sktime/forecasting/compose/_reduce.py]: We have three strategies. `direct`, `recursive` and (in this PR) `multioutput`. 

Markus was interested in knowing what error message arises if person X tries to use the multioutput strategy with a regressor R that does not support multioutput. You then suggested that X could wrap the regressor R using sklearn's `MultiOutputRegressor`, so you'd get something like: `MultioutputReductionForecaster(MultiOutputRegressor(R),...)`

I then said that you do not need this multitoutput as we can just use direct:
`DirectionReductionForecaster(R,...)`

Comment (fkiraly) [reply sktime/forecasting/compose/_reduce.py]: Get it now.
So, I'm saying: in that case (regressor does not have multi-output support), I'd *still* wrap it in `MultiOutputRegressor`.
Why: because then overall behaviour is consistent and does not raise an error when using the `multioutput` strategy

Even if in that special case `multioutput` would give you the same behaviour as `direct`, it's fine - since it's not the case in general (just in that case). Both ways get you the expected result rather than an error message which wouldn't be clear why it is raised, if the user just starts with specifying either `multioutput` or `direct`.

Comment (mloning) [reply sktime/forecasting/compose/_reduce.py]: Okay, I like that solution too. But the problem still remains that we need to check if the passed regressor support multioutput natively or not. And I would add a warning if it doesn't, since we change the input and wrap it automatically in the `MultioutputRegressor` we should let the user know about it.

**Comment by mloning** [comment]
@Lovkush-A can you ping me again whenever you think this is good to go?

File: sktime/forecasting/compose/_reduce.py
Comment (Lovkush-A) [reply sktime/forecasting/compose/_reduce.py]: My opinion is that using a non-multitoutput regressor in the multioutput regression should raise an error (as it currently will). Maybe there is a principle of software design I do not know about, but I think a user should be incentivised to use the tools properly. If they want direct strategy, they should choose it, if they want multioutput, they should choose an appropriate regressor.

**Comment by Lovkush-A** [comment]
> @Lovkush-A can you ping me again whenever you think this is good to go?

@mloning I am fine with it.

File: sktime/forecasting/compose/_reduce.py
Comment (mloning) [reply sktime/forecasting/compose/_reduce.py]: I agree @Lovkush-A, that's why I think raising the warning may be a compromise. I suggest to leave it as it is for now, merge the PR and wait for some user feedback, that also saves us from having to check the estimator type.

File: sktime/forecasting/compose/_reduce.py
Comment (mloning) [inline sktime/forecasting/compose/_reduce.py:63]: Once we have merged #660, we need to remove the `or update_params` here and update the error message.

**Review by mloning — COMMENTED** [review]
Okay thanks @Lovkush-A - this is good to go with one caveat below depending another PR which I like to merge first.

File: sktime/forecasting/compose/_reduce.py
Comment (mloning) [reply sktime/forecasting/compose/_reduce.py:63]: Now merged

**Review by mloning — APPROVED** [review]

**Comment by mloning** [comment]
Thanks @Lovkush-A - now merged! 🎉

**Pull request merged**
